### PR TITLE
PyCode "Unparse": Added support for slice objects in subscriptions

### PR DESCRIPTION
**Pull request by MatrixEditor**
Subject: Fix documentation issue when slice objects are used in subscriptions

### Feature or Bugfix
- Feature

### Purpose
- Added function to handle parsed slice objects

### Detail
- For instance, consider the following class definition:
  ```python
  class Foo: ...

  class Bar(BaseType):
      #: doc comment
      field1: int
      field2: Foo[1:2:3]
      """doc comment"""
  ```
  Internally, sphinx will raise an exception while parsing this file:
  ```
  ...
  File "venv-3.12/lib/python3.12/site-packages/sphinx/pycode/ast.py", line 170, in visit_Subscript
    return f"{self.visit(node.value)}[{self.visit(node.slice)}]"
                                       ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/ast.py", line 407, in visit
    return visitor(node)
           ^^^^^^^^^^^^^
  File "venv-3.12/lib/python3.12/site-packages/sphinx/pycode/ast.py", line 188, in generic_visit
    raise NotImplementedError('Unable to parse %s object' % type(node).__name__)
  NotImplementedError: Unable to parse Slice object
  ```
- *The implemented function does not alter existing code*

### Relates

- #10817

### CODE REVIEW

**Comment by picnixz** [comment]
Can you add tests and a CHANGES entry?

**Review by picnixz — CHANGES_REQUESTED** [review]
Can you add tests and a CHANGES please?

File: sphinx/pycode/ast.py [hunk 1]
```diff
@@ -187,3 +187,9 @@ def visit_Tuple(self, node: ast.Tuple) -> str:
 
     def generic_visit(self, node: ast.AST) -> NoReturn:
         raise NotImplementedError('Unable to parse %s object' % type(node).__name__)
+
+    def visit_Slice(self, node: ast.Slice) -> str:
```
Comment (picnixz) [inline sphinx/pycode/ast.py]: Can you put the method above in alphabetical order ? (I think the `visit_*` methods are alphabetically ordered).

File: tests/test_pycode/test_pycode_ast.py [hunk 2]
```diff
@@ -51,6 +51,8 @@
      "lambda x=0, /, y=1, *args, z, **kwargs: ..."),  # posonlyargs
     ("0x1234", "0x1234"),                       # Constant
     ("1_000_000", "1_000_000"),                 # Constant
+    ("Tuple[:,:]", "Tuple[::, ::]"),            # Index, Subscript, Slice
```
Comment (picnixz) [inline tests/test_pycode/test_pycode_ast.py]: `[::,::]` feel weird and incorrect to me. For the entirely empty slice, I think `:` should be used instead (i.e., when node.lower, node.upper and node.step are not set).

Comment (MatrixEditor) [reply tests/test_pycode/test_pycode_ast.py]: You're right - fixing that in a bit.

File: tests/test_pycode/test_pycode_ast.py [hunk 3]
```diff
@@ -51,6 +51,9 @@
      "lambda x=0, /, y=1, *args, z, **kwargs: ..."),  # posonlyargs
     ("0x1234", "0x1234"),                       # Constant
     ("1_000_000", "1_000_000"),                 # Constant
+    ("Tuple[:,:]", "Tuple[:, :]"),              # Index, Subscript, 2x Slice
```
Comment (picnixz) [inline tests/test_pycode/test_pycode_ast.py:54]: I'm a bit nitpicky, but do you think it's prefereable to have a space after `,` ? Also, could you perhaps add some tests using numpy extended syntax where you have commas + slices?

Comment (MatrixEditor) [reply tests/test_pycode/test_pycode_ast.py:54]: No problem. Maybe the example isn't the best one, but I still think that space is okay (the space would be added automatically when reformatting a document e.g. using black). *The space is not added by the `visit_Slice`* function.

Comment (picnixz) [reply tests/test_pycode/test_pycode_ast.py:54]: Good for me then !

File: tests/test_pycode/test_pycode_ast.py [hunk 4]
```diff
@@ -51,6 +51,13 @@
...
+    ("Tuple[1:2]", "Tuple[1:2]"),               # Index, Subscript, Slice(no-step)
+    ("Tuple[1:2:3]", "Tuple[1:2:3]"),           # Index, Subscript, Slice
+    ("x[:, np.newaxis, :, :]",
+     "x[:, np.newaxis, :, :]"),                 # Index, Subscript, numpy extended syntax
+    ("y[:, 1:3][np.array([0, 2, 4]), :]",
+     "y[:, 1:3][np.array([0, 2, 4]), :]"),       # Index, 2x Subscript, numpy extended syntax
```
Comment (picnixz) [inline tests/test_pycode/test_pycode_ast.py:60]: Just by looking at all those identity checks, I think we should have used a sentinel value, say ``IDENTICAL``, to indicate that the output is the same as the input.  This is an easy follow-up PR though.

**Review by picnixz — APPROVED** [review]

File: CHANGES.rst [hunk 5]
```diff
@@ -27,6 +27,8 @@ Features added
 * #11892: Improved performance when resolving cross references in cpp domain.
   Patch by Rouslan Korneychuk.
 
+* #11981: pycode "UnparseVisitor": Added support for slice objects in subscriptions
```
Comment (picnixz) [inline CHANGES.rst]:
```suggestion
* #11981: Improve rendering of signatures using ``slice`` syntax,
  e.g., ``def foo(arg: np.float64[:,:]) -> None: ...``.
```

**Review by picnixz — COMMENTED** [review]
Actually, can you check that https://github.com/sphinx-doc/sphinx/issues/10817 is indeed solved?

**Comment by MatrixEditor** [comment]
Yes, but that will take me a few moments.

**Comment by picnixz** [comment]
Don't worry, you're fast enough (it also allows me to work on something else in the meantime)

**Comment by MatrixEditor** [comment]
Yes, #10817 is indeed fixed:

temp/\_\_init\_\_.py
```python
Ac = "float64[:, :]"


def myfunc(arg: Ac):
    """Some docs"""
    ...
```

index.rst
```rst
Welcome to temp's documentation!
================================

.. autofunction:: temp.myfunc


Indices and tables
==================

* :ref:`genindex`
* :ref:`modindex`
* :ref:`search`
```

Results in:
image

**Comment by picnixz** [comment]
Is it possible to include this one as a test as well?

**Comment by picnixz** [comment]
Thank you!

**Pull request merged**